File: src/clj/curatify/spotify.clj
(ns curatify.spotify
  (:require [ring.util.codec :refer [form-encode]]
            [curatify.config :refer [env]]
            [clj-http.client :as client]
            [clojure.tools.logging :as log]
            [clojure.string :refer [join]]))

(defn client-id [] (-> env :spotify :client-id))
(defn client-secret [] (-> env :spotify :client-secret))
(defn redirect-uri [] (-> env :spotify :redirect-uri))


(defn api-req
  ([method url access] (api-req method url access {}))
  ([method url access opts]
   (method url (conj opts {:as :json :headers {"Authorization" (str "Bearer " access)}}))))


(def accounts "https://accounts.spotify.com")
(def api "https://api.spotify.com/v1")
(def scope "user-read-email user-library-read user-library-modify streaming user-read-birthdate user-read-private user-modify-playback-state")


(defn auth-url []
  (str accounts "/authorize?"
       (form-encode {:client_id (client-id)
                     :redirect_uri (redirect-uri)
                     :response_type "code"
                     :scope scope})))


(defn enrich-token [token]
  (assoc token :expires-at (+ (:expires_in token)
                              (quot (System/currentTimeMillis) 1000))))

(defn get-token [code]
  (enrich-token (:body (client/post (str accounts "/api/token")
                                    {:form-params {:grant_type "authorization_code"
                                                   :code code
                                                   :redirect_uri (redirect-uri)
                                                   :client_id (client-id)
                                                   :client_secret (client-secret)}
                                     :as :json}))))


(defn get-generic-token []
  (:body (client/post (str accounts "/api/token")
                    {:form-params {:grant_type "client_credentials"
                                   :client_id (client-id)
                                   :client_secret (client-secret)}
                     :as :json})))


(defn refresh-token [{token :refresh_token}]
  (enrich-token (:body (client/post (str accounts "/api/token")
                                {:form-params {:grant_type "refresh_token"
                                               :client_id (client-id)
                                               :client_secret (client-secret)
                                               :refresh_token token}
                                 :as :json}))))


(defn me [{access :access_token}]
  (:body (api-req client/get (str api "/me") access)))


(defn paginate [access-token endpoint]
  (loop [next-page (str api endpoint)
         acc []]
    (if (nil? next-page)
      acc
      (let [{{:keys [items next]} :body} (api-req client/get next-page access-token)]
        (recur next
               (into [] (concat acc items)))))))


(defn me-playlists [{access :access_token}]
  (paginate access "/me/playlists"))


(defn playlist-tracks [playlist-id {access :access_token}]
  (paginate access (str "/playlists/" playlist-id "/tracks")))


(defn artists [artist-ids]
  (let [{access :access_token} (get-generic-token)]
    (:artists (:body (api-req client/get
                          (str api "/artists")
                          access
                          {:query-params {:ids (join "," artist-ids)}})))))

File: lib/spotify.ts
import { PlaylistTrack, Track } from "./../server/data/tracks";
import { User } from "./../server/data/users";
import { Artist } from "./../server/data/artists";
import axios, { AxiosInstance } from "axios";
import axiosRetry from "axios-retry";
import { stringify } from "querystring";
import camelcaseKeys from "camelcase-keys";
import { Token } from "../server/data/users";
import { Playlist } from "../server/data/playlists";

interface WrappedPlaylistTrack {
  track: Track;
  id: string;
}

export class SpotifyClient {
  static authBase = "https://accounts.spotify.com";
  static auth = axios.create({ baseURL: SpotifyClient.authBase });
  static scope = [
    "user-read-email",
    "user-library-read",
    "user-library-modify",
    "streaming",
    "user-read-birthdate",
    "user-read-private",
    "user-modify-playback-state",
    "playlist-modify-private",
    "playlist-modify-public"
  ].join(" ");

  token: Token;
  api: AxiosInstance;

  constructor(token: Token) {
    this.token = token;
    this.api = axios.create({
      baseURL: "https://api.spotify.com/v1",
      headers: { Authorization: `Bearer ${token.accessToken}` }
    });
    axiosRetry(this.api, { retries: 3 });
  }

  static enrichToken(token: Token): Token {
    const camelToken = camelcaseKeys(token);
    return {
      expiresAt: new Date().getTime() / 1000 + camelToken.expiresIn,
      ...camelToken
    };
  }

  static async getToken(): Promise<Token> {
    const { data } = await this.auth.post(
      "/api/token",
      stringify({
        client_id: process.env.SPOTIFY_ID,
        grant_type: "client_credentials",
        client_secret: process.env.SPOTIFY_SECRET
      })
    );
    return this.enrichToken(data);
  }

  async artists(artistIds: string[]): Promise<Artist[]> {
    const {
      data: { artists }
    } = await this.api.get("/artists", {
      params: { ids: artistIds.filter(id => !!id).join(",") }
    });
    return artists;
  }
}

export class SpotifyUserClient extends SpotifyClient {
  static authUrl(): string {
    return `${SpotifyUserClient.authBase}/authorize?${stringify({
      scope: SpotifyUserClient.scope,
      client_id: process.env.SPOTIFY_ID,
      redirect_uri: process.env.CALLBACK_URL,
      response_type: "code"
    })}`;
  }

  static async getToken(code?: string): Promise<Token> {
    const { data } = await SpotifyUserClient.auth.post(
      "/api/token",
      stringify({
        client_id: process.env.SPOTIFY_ID,
        redirect_uri: process.env.CALLBACK_URL,
        grant_type: "authorization_code",
        client_secret: process.env.SPOTIFY_SECRET,
        code
      })
    );
    return super.enrichToken(data);
  }

  async refreshToken(): Promise<Token> {
    const { data } = await SpotifyUserClient.auth.post(
      "/api/token",
      stringify({
        client_id: process.env.SPOTIFY_ID,
        grant_type: "refresh_token",
        client_secret: process.env.SPOTIFY_SECRET,
        refresh_token: this.token.refreshToken
      })
    );
    return SpotifyClient.enrichToken({ ...this.token, ...data });
  }

  async me(): Promise<User> {
    const { data } = await this.api.get("/me");
    return data;
  }

  async *mePlaylists(): AsyncIterable<Playlist[]> {
    let next = "/me/playlists";
    while (next) {
      const { data } = await this.api.get(next);
      next = data.next;
      yield data.items;
    }
  }

  async *playlistTracks(
    playlistId: string
  ): AsyncIterable<WrappedPlaylistTrack[]> {
    let next = `/playlists/${playlistId}/tracks`;
    while (next) {
      const { data } = await this.api.get(next, { params: { limit: 100 } });
      next = data.next;
      yield data.items;
    }
  }

  async play(
    uris: string[],
    deviceId: string,
    position_ms?: string
  ): Promise<void> {
    await this.api.put(`/me/player/play?device_id=${deviceId}`, {
      uris,
      position_ms
    });
  }

  async pause(deviceId: string): Promise<void> {
    await this.api.put(`/me/player/pause?device_id=${deviceId}`);
  }

  async addTrackToPlaylist(trackId: string, playlistId: string): Promise<void> {
    await this.api.post(`/playlists/${playlistId}/tracks`, {
      uris: [`spotify:track:${trackId}`]
    });
  }
}
